File: examples/test1/account.lua
--
-- Lua 5.2 & Luna Five: Account Example
-- Date: 16 May 2013
-- Website: http://blog.p86.dk
--

-- Override Lua Print()
local print = function (x)
  print("Lua: "..x)
end

local a1 = Account(100)
print("Created Account 1 with $"..a1:getBalance().." balance")

cash = 50
a1.deposit(cash)
print("Account 1: Depositing $"..cash)

print("Account 1: New Balance is $"..a1:getBalance())

cash = 25
a1.withdraw(cash)
print("Account 1: Withdrawing $"..cash)

print("Account 1: New Balance is $"..a1:getBalance())


-- Open a 2nd Account
local a2 = Account(1000)
print("Created Account 2 with $"..a2:getBalance().." balance")

cash = 500
a2.deposit(cash)
print("Account 2: Depositing $"..cash)

print("Account 2: New Balance is $"..a2:getBalance())

cash = 250
a2.withdraw(cash)
print("Account 2: Withdrawing $"..cash)

print("Account 2: New Balance is $"..a2:getBalance())


-- Check the 1st Account Again
print("Account 1: New Balance is $"..a1:getBalance())
cash = 25
a1.withdraw(cash)
print("Account 1: Withdrawing $"..cash)

print("Account 1: New Balance is $"..a1:getBalance())

local a3 = Account(100)
a3.X = 202;
print("X equals " .. a3.X)

print("End of Script")

File: examples/test1/Account.cpp

#include "Account.h"

#define method(class, name) {#name, &class::name}

// name to use to address the class in Lua
const char Account::className[] = "Account";

// list of class properties that one can set/get from Lua
const Luna<Account>::PropertyType Account::properties[] = {
  {"X", &Account::getX, &Account::setX},
  {0,0,0}
};

// list of class methods to make available in Lua
const Luna<Account>::FunctionType Account::methods[] = {
  method(Account, deposit),
  method(Account, withdraw),
  method(Account, getBalance),
  {0,0}
};

Account::Account(lua_State* l)
{
   balance = lua_tointeger(l, 1);
   x = 0;
}

Account::~Account()
{
   printf("C++: Deleted Account (%p)\n", this);
}

int Account::setX(lua_State* l)
{
   x = lua_tointeger(l, 1);
   return 0;
}

int Account::getX(lua_State* l)
{
   lua_pushnumber(l, x);
   return 1;
}

int Account::deposit(lua_State* l)
{
   balance += lua_tointeger(l, 1);
   return 0;
}

int Account::withdraw(lua_State* l)
{
   balance -= lua_tointeger(l, 1);
   return 0;
}

int Account::getBalance(lua_State* l)
{
   lua_pushnumber(l, balance);
   return 1;
}
